gh-143120: pixi builds for free-threading and TSAN

crusaderky opened the pull request:
Follow-up to #142469

- Add pixi infrastructure for
  - `freethreading`: compiles with `--disable-gil`
  - `tsan-freethreading`: compiles with `--disable-gil --with-thread-sanitizer`
- Refactor recipes:
  - Make `recipe.yaml` and `pixi.toml` all identical symlinks
  - Remove clang-19 since we are not building the experimental jit
  - Remove ld_impl because we are not hacking the makefiles
  - Add site_packages_path
- Add support for MacOS Intel (specifically `macos-15-intel` github runners)
- Signal support for Linux ARM (it already worked but was insofar untested)
- Remove ASAN/TSAN DLLs hack on MacOS

Tested in CI on https://github.com/crusaderky/hdf5-pixi

<!-- gh-issue-number: gh-143120 -->
* Issue: gh-143120
<!-- /gh-issue-number -->

crusaderky committed:
pixi builds for free-threading and tsan

crusaderky commented on Tools/pixi-packages/free-threading/recipe.yaml:
```diff
@@ -0,0 +1,63 @@
+  script:
+    file: ../build.sh
+    env:
+      PYTHON_VARIANT: "free-threading"
```
copy-paste of `asan/recipe.yaml`, except this one line

crusaderky commented on Tools/pixi-packages/tsan-free-threading/pixi.toml:
copy-paste of `asan/pixi.toml`

crusaderky commented on Tools/pixi-packages/tsan/recipe.yaml:
```diff
@@ -0,0 +1,63 @@
+  script:
+    file: ../build.sh
+    env:
+      PYTHON_VARIANT: "tsan"
```
copy-paste of `asan/recipe.yaml`, except this one line

ngoldbaum commented on Tools/pixi-packages/README.md:
```diff
@@ -12,15 +12,17 @@ python = { git = "https://github.com/python/cpython", subdirectory = "Tools/pixi
-- `asan`: ASan-instrumented build with `PYTHON_ASAN=1`
+- `free-threading`
+- `asan`: ASan-instrumented build
+- `tsan`: free-threading, TSan-instrumented build
```
Should probably be called `tsan-free-threading`. There should probably also be an `asan-free-threading`. @lucascolley is there a way to avoid the combinatorial explosion of variants here? Ideally you'd be able to somehow combine these but I think that's probably not tenable at the moment.

crusaderky replied:
I think the answer to this is no because of https://github.com/prefix-dev/pixi/issues/4599
Barring the above, we could write a template + generation script and then commit the output to git? Very ugly if you ask me.

lucascolley replied:
Yes, I would vote for keeping things simple but verbose at the minute until there is a proper solution upstream

crusaderky replied:
Renamed to `tsan-free-threading`

crusaderky replied:
Renamed to tsan-freethreading in line with cpython recipes

ngoldbaum commented on Tools/pixi-packages/build.sh:
```diff
@@ -1,11 +1,19 @@
     echo "BUILD TYPE: ASAN"
     BUILD_DIR="../build_asan"
     CONFIGURE_EXTRA="--with-address-sanitizer"
-    export PYTHON_ASAN="1"
```
why delete this?

crusaderky replied:
I did not find it to be used anywhere. @lucascolley any insight?

lucascolley replied:
I think it can be removed, yes. I forgot halfway through whether it did anything on the CPython side

ngoldbaum reviewed (commented):
I opened an issue which is needed to make this mergeable. I'm also playing with a NumPy recipe based on this.

crusaderky committed:
Rename to tsan-free-threading

crusaderky committed:
Add libclang_rt hack

crusaderky commented on Tools/pixi-packages/build.sh:
```diff
@@ -33,5 +41,6 @@ ln -sf "${PREFIX}/bin/python3" "${PREFIX}/bin/python"
 if [[ ${OSTYPE} == "darwin"* ]]; then
-    cp "${BUILD_PREFIX}/lib/clang/21/lib/darwin/libclang_rt.asan_osx_dynamic.dylib" "${PREFIX}/lib/libclang_rt.asan_osx_dynamic.dylib"
+    cp "${BUILD_PREFIX}/lib/clang/21/lib/darwin/libclang_rt.asan_osx_dynamic.dylib" "${PREFIX}/lib/"
+    cp "${BUILD_PREFIX}/lib/clang/21/lib/darwin/libclang_rt.tsan_osx_dynamic.dylib" "${PREFIX}/lib/"
```
I don't have straightforward means to test this. @lucascolley / @ngoldbaum could you give it a go?

crusaderky replied:
The build dependencies list `{compiler('c')}`, which I believe translates to clang-21, plus and explicit reference to `clang-19`. Here we're referencing clang-21.
@lucascolley is this deliberate?

lucascolley replied:
maybe we can drop the `clang-19`

happy to give this a go but would be easier once there is a NumPy build using this.

ngoldbaum replied:
I'm able to get past the rpath errors I was seeing before on my Mac using this hack. Thanks!

crusaderky committed:
Add troubleshooting for tsan

isuruf committed:
Add python_abi output and refactor

- Remove clang-19 since we are not building the experimental jit
- Remove ld_impl because we are not hacking the makefiles
- Add site_packages_path
- Make recipe.yaml identical with just the variant name difference
- Add a generate.sh to generate recipe.yaml from default variant
- Add a runtime requirement of libsanitizer for gcc

isuruf committed:
fix typo

isuruf committed:
Use single recipe.yaml duplicated across directories

isuruf committed:
use symlinks

isuruf committed:
remove generate.sh

crusaderky committed:
Merge remote-tracking branch 'origin/main' into tsan

crusaderky committed:
Fix

crusaderky committed:
Merge remote-tracking branch 'upstream/main' into tsan

crusaderky committed:
Fix tsan

lucascolley commented on Tools/pixi-packages/asan/pixi.toml:
```diff
@@ -3,6 +3,11 @@ channels = ["https://prefix.dev/conda-forge"]
+[workspace.build-variants]
+variant = ["asan"]
+abi_tag = ["cp315"]
+version = ["3.15"]
```
can this be deleted now?

crusaderky committed:
switch ABI tag for asan/tsan

crusaderky committed:
git cleanup 1/2

crusaderky committed:
git cleanup 2/2

lucascolley reviewed (commented):
could we update the README to point to the new blockers for using a single manifest and recipe per https://github.com/python/cpython/pull/142872#issuecomment-3750130564 ?

crusaderky committed:
README

crusaderky committed:
Merge remote-tracking branch 'upstream/main' into tsan

crusaderky commented on Tools/pixi-packages/build.sh:
```diff
@@ -23,6 +29,7 @@ else
         --oldincludedir="${BUILD_PREFIX}/${HOST}/sysroot/usr/include" \
         --enable-shared \
         --srcdir="${SRC_DIR}" \
+        --with-system-expat \
```
@isuruf not sure what this is about?

isuruf replied:
Since we install `expat` from conda, this tells cpython to use that.

isuruf committed:
Use libcompiler-rt and remove hack

isuruf committed:
Use libcompiler-rt and remove hack

ngoldbaum commented on Tools/pixi-packages/build.sh:
```diff
@@ -33,5 +40,6 @@ ln -sf "${PREFIX}/bin/python3" "${PREFIX}/bin/python"
 if [[ ${OSTYPE} == "darwin"* ]]; then
-    cp "${BUILD_PREFIX}/lib/clang/21/lib/darwin/libclang_rt.asan_osx_dynamic.dylib" "${PREFIX}/lib/libclang_rt.asan_osx_dynamic.dylib"
+    cp "${BUILD_PREFIX}/lib/clang/21/lib/darwin/libclang_rt.asan_osx_dynamic.dylib" "${PREFIX}/lib/"
+    cp "${BUILD_PREFIX}/lib/clang/21/lib/darwin/libclang_rt.tsan_osx_dynamic.dylib" "${PREFIX}/lib/"
```
See https://github.com/crusaderky/cpython/pull/2 for a better way to do this

isuruf committed:
use at least macos 10.12

crusaderky committed:
Merge remote-tracking branch 'upstream/main' into tsan

crusaderky committed:
note

lucascolley commented on Tools/pixi-packages/README.md:
```diff
@@ -1,36 +1,59 @@
+  and [pixi#5248](https://github.com/prefix-dev/pixi/issues/5248)
+- A workaround can be removed from the build script once
+  [rattler-build#2012](https://github.com/prefix-dev/rattler-build/issues/2012) is
+  resolved
```
```suggestion
```

isuruf committed:
Fix CPPFLAGS too

crusaderky committed:
Merge remote-tracking branch 'upstream/main' into tsan

crusaderky committed:
Merge remote-tracking branch 'upstream/main' into tsan

crusaderky committed:
Update Tools/pixi-packages/README.md

Co-authored-by: Lucas Colley <lucas.colley8@gmail.com>

crusaderky committed:
Update platforms

crusaderky committed:
Merge remote-tracking branch 'isuruf/libcompiler-rt' into tsan

crusaderky committed:
Merge branch 'main' into tsan

crusaderky committed:
Revert "Use libcompiler-rt and remove hack"

This reverts commit ad6c74054b3d5bef02bd262ce1b9648e0aad3b09.

crusaderky committed:
Revert "Update Tools/pixi-packages/README.md"

This reverts commit 26492862e5bd2d261a23b912cd1549c0cb59dd10.

crusaderky committed:
Reapply "Update Tools/pixi-packages/README.md"

This reverts commit cddc9336f8ace230db2fdbe398696b931019f2d0.

crusaderky committed:
Reapply "Use libcompiler-rt and remove hack"

This reverts commit 1ae77a51883dc0b2db7e961ddab31c4963e3f979.

crusaderky committed:
There is no c_compiler variable

crusaderky committed:
Don't use system-wide libasan on Linux

crusaderky committed:
Merge remote-tracking branch 'upstream/main' into tsan

crusaderky committed:
Don't use symlinks 1/2

crusaderky committed:
Don't use symlinks 2/2

kumaraditya303 reviewed (approved):
I tested the tsan+freethreading pixi config with the following and it worked:
```toml
[workspace]
channels = ["conda-forge"]
name = "cpython-pixi-test"
platforms = ["linux-64"]
preview = ["pixi-build"]

[dependencies]
python = { git = "https://github.com/crusaderky/cpython", rev = "tsan", subdirectory = "Tools/pixi-packages/tsan-freethreading" }
```

I'll merge this by tomorrow if there are no objections.

kumaraditya303 committed:
Merge branch 'main' into tsan

hugovk commented on Tools/pixi-packages/README.md:
```diff
@@ -1,36 +1,56 @@
-- Using a single `pixi.toml` and `recipe.yaml` for all package variants is blocked on https://github.com/prefix-dev/pixi/issues/4599
-- A workaround can be removed from the build script once https://github.com/prefix-dev/rattler-build/issues/2012 is resolved
+- Using a single `pixi.toml` and `recipe.yaml` for all package variants is blocked on
+  [pixi-build-backends#532](https://github.com/prefix-dev/pixi-build-backends/pull/532)
```
No need to update the `README.md` right now, but noting this repo has been archived and the new PR is at https://github.com/prefix-dev/pixi/pull/5364.

hugovk commented on Tools/pixi-packages/README.md:
````diff
@@ -1,36 +1,56 @@
+```
+To fix it, try reducing `mmap_rnd_bits`:
+
+```bash
````
Tip for next time: `bash` for when there's only commands, and `console` for when there's commands with output. Compare:

```bash
$ sudo sysctl vm.mmap_rnd_bits
vm.mmap_rnd_bits = 32  # too high for TSan
$ sudo sysctl vm.mmap_rnd_bits=28  # reduce it
vm.mmap_rnd_bits = 28
```


```console
$ sudo sysctl vm.mmap_rnd_bits
vm.mmap_rnd_bits = 32  # too high for TSan
$ sudo sysctl vm.mmap_rnd_bits=28  # reduce it
vm.mmap_rnd_bits = 28
```
